nuxeo-mongodb-apb/templates/statefulset.yml.j2

```diff
@@ -0,0 +1,304 @@
+              mountPath: /data/db
+      containers:
+        - name: {{ apb_name }}
+          image: "{{ image_repository }}:{{ image_tag }}"
+          imagePullPolicy: "{{ image_pullPolicy }}"
+{% if extraVars is defined and (extraVars | from_yaml) %}
+          env:
+{{ extraVars | from_yaml | to_nice_yaml(indent=2) | indent(12, true) }}
+{% endif %}
+          ports:
+            - name: peer
+              containerPort: 27017
+          resources:
+{% if resources is defined and (resources | from_yaml) %}
+{{ resources | from_yaml | to_nice_yaml(indent=2) | indent(12, true) }}
+{% endif %}
+          command:
+            - mongod
+          args:
+            - --config=/data/configdb/mongod.conf
+            - --dbpath=/data/db
+            - --replSet={{ replicaSetName }}
+            - --port={{ port }}
+            - --bind_ip
+            - "0.0.0.0"
+{% if auth_enabled %}
+            - --auth
+{% endif %}
+{% if tls_enabled %}
+            - --clusterAuthMode
+            - x509
+            - --sslMode
+            - requireSSL
+            - --sslCAFile=/data/configdb/tls.crt
+            - --sslPEMKeyFile=/work-dir/mongo.pem
+{% endif %}
+          livenessProbe:
+            exec:
+              command:
+                - mongo
+{% if tls_enabled %}
+                - --ssl
+                - --sslCAFile=/data/configdb/tls.crt
+                - --sslPEMKeyFile=/work-dir/mongo.pem
+{% endif %}
+                - --eval
+                - "db.adminCommand('ping')"
+            initialDelaySeconds: {{ livenessProbe_initialDelaySeconds }}
+            timeoutSeconds: {{ livenessProbe_timeoutSeconds }}
+            failureThreshold: {{ livenessProbe_failureThreshold }}
+            periodSeconds: {{ livenessProbe_periodSeconds }}
+            successThreshold: {{ livenessProbe_successThreshold }}
+          readinessProbe:
+            exec:
+              command:
+                - mongo
+{% if tls_enabled %}
+                - --ssl
+                - --sslCAFile=/data/configdb/tls.crt
+                - --sslPEMKeyFile=/work-dir/mongo.pem
+{% endif %}
+                - --eval
+                - "db.adminCommand('ping')"
+            initialDelaySeconds: {{ readinessProbe_initialDelaySeconds }}
+            timeoutSeconds: {{ readinessProbe_timeoutSeconds }}
+            failureThreshold: {{ readinessProbe_failureThreshold }}
+            periodSeconds: {{ readinessProbe_periodSeconds }}
+            successThreshold: {{ readinessProbe_successThreshold }}
+          volumeMounts:
+            - name: datadir
+              mountPath: /data/db
+            - name: configdir
+              mountPath: /data/configdb
+            - name: workdir
+              mountPath: /work-dir
+{% if nodeSelector is defined and (nodeSelector | from_yaml) %}
+      nodeSelector:
+{{ nodeSelector | from_yaml | to_nice_yaml(indent=2) | indent(8, true) }}
+{% endif %}
+{% if affinity is defined and (affinity | from_yaml) %}
+      affinity:
```

Comment:
I know this is derived from the helm charts, but I would rather put something more opinionated here: affinity is a bit hard to express with a simple parameter. We should at least put some anti-affinity on what we know:

```
podAntiAffinity:
    preferredDuringSchedulingIgnoredDuringExecution:
    - weight: 1
      podAffinityTerm:
        labelSelector:
          matchExpressions:
          - key: "app"
            operator: In
            values:
            - "{{ name }}"
          - key: "apb_plan_id"
            values:
            - "{{ apb_plan_id }}"
          - key: "component"
            operator: In
            values:
            - "{{ component }}"
        topologyKey: "{{ topology_key }}"
```

`toplogyKey` could be a choice from:

 - `kubernetes.io/hostname`
 - `failure-domain.beta.kubernetes.io/zone`
 - `failure-domain.beta.kubernetes.io/region`
 - `beta.kubernetes.io/instance-type`
 - `beta.kubernetes.io/os`
 - `beta.kubernetes.io/arch`

and let affinity based on pod labels being configurable in YAML. This would at least help a lot how to configure the 20/80 cases.

WDYT?

Reply:
I suspect production deployments will use the command-line `svcat` utility and specify the parameters using a json parameters file. In this case, it is easy to include an affinity definition similar to the one you proposed. What would you think about including the above affinity definition in the (*)secure-params.json files rather than hard-wiring it into the apb?

Reply:
I think we are pursuing different goals here. Giving full control is ok for production like deployment where we can expect people know what they are doing.

Another goal is also to have a painless way of deploying a rather good MongoDB deployment that suits Nuxeo using the webconsole. 

Should we add  `useDefaultAntiAffinity` and `defautTopologyKey`  parameter to the APB ?

Reply:
Yes, I think our goals are a bit different here.

I like your idea about enhancing my patch to include the two new affinity-related parameters. Since the current code that I am patching does not have this feature, would you consider applying my revised patch without this feature? Then, we could treat the proposed default anti-affinity feature as an independent, follow-on enhancement. WDYT? I would like to get my patch incorporated ASAP so that we can begin testing it in our internal environment.

Reply:
👍